utils/test_results.py

```diff
@@ -1,73 +1,295 @@
-    function = "distinct"
+@dataclass
+class TestResultsRow:
+    # the order here must match the order of the fields in the query
+    name: str
+    test_id: str
+    failure_rate: float
+    flake_rate: float
+    updated_at: dt.datetime
+    avg_duration: float
+    total_fail_count: int
+    total_flaky_fail_count: int
+    total_pass_count: int
+    total_skip_count: int
+    commits_where_fail: int
+    last_duration: float
 
 
-class Array(Aggregate):
-    function = "array"
+@dataclass
+class Connection:
+    edges: list[dict[str, str | TestResultsRow]]
+    page_info: dict
+    total_count: int
 
 
-class GENERATE_TEST_RESULT_PARAM:
-    FLAKY = "flaky"
-    FAILED = "failed"
-    SLOWEST = "slowest"
-    SKIPPED = "skipped"
+def convert_tuple_else_none(
+    value: set[str] | list[str] | None,
+) -> tuple[str, ...] | None:
+    return tuple(value) if value else None
+
+
+@dataclass
+class CursorValue:
+    ordered_value: str
+    name: str
+
+
+def decode_cursor(value: str | None) -> CursorValue | None:
+    if value is None:
+        return None
+
+    split_cursor = b64decode(value.encode("ascii")).decode("utf-8").split(DELIMITER)
+    return CursorValue(
+        ordered_value=split_cursor[0],
+        name=split_cursor[1],
+    )
+
+
+def encode_cursor(row: TestResultsRow, ordering: str) -> str:
+    return b64encode(
+        DELIMITER.join([str(getattr(row, ordering)), str(row.name)]).encode("utf-8")
+    ).decode("ascii")
+
+
+def validate(
+    interval_num_days: int,
+    ordering: str,
+    ordering_direction: str,
+    after: str | None,
+    before: str | None,
+    first: int | None,
+    last: int | None,
+) -> ValidationError | None:
+    if interval_num_days not in {1, 7, 30}:
+        return ValidationError(f"Invalid interval: {interval_num_days}")
+
+    if ordering_direction not in {"ASC", "DESC"}:
+        return ValidationError(f"Invalid ordering direction: {ordering_direction}")
+
+    if ordering not in {
+        "name",
+        "computed_name",
+        "avg_duration",
+        "failure_rate",
+        "flake_rate",
+        "commits_where_fail",
+        "last_duration",
+    }:
+        return ValidationError(f"Invalid ordering field: {ordering}")
+
+    if first is not None and last is not None:
+        return ValidationError("First and last can not be used at the same time")
+
+    if after is not None and before is not None:
+        return ValidationError("After and before can not be used at the same time")
+
+    return None
+
+
+def generate_base_query(
+    repoid: int,
+    ordering: str,
+    ordering_direction: str,
+    should_reverse: bool,
+    branch: str | None,
+    interval_num_days: int,
+    testsuites: list[str] | None = None,
+    term: str | None = None,
+    test_ids: set[str] | None = None,
+) -> TestResultsQuery:
+    term_filter = f"%{term}%" if term else None
+
+    if should_reverse:
+        ordering_direction = "DESC" if ordering_direction == "ASC" else "ASC"
```

Comment:
is this a typo?

Reply:
nope, we want to reverse the ordering direction when we get `last` passed to the query

Reply:
oh I see, yeah that makes sense